torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -287,6 +287,45 @@ def module(self) -> FlattenParamsWrapper:
         assert isinstance(self._fsdp_wrapped_module, FlattenParamsWrapper)
         return self._fsdp_wrapped_module
 
+    def fsdp_modules(self) -> List["FullyShardedDataParallel"]:
+        """
+        Helper function to return all nested FSDP instances, including self.
+        """
+        fsdp_modules = []
+        for module in self.modules():
+            if isinstance(module, FullyShardedDataParallel):
+                fsdp_modules.append(module)
+
+        return fsdp_modules
+
+    def apply(self, fn: Callable[[nn.Module], None]) -> "FullyShardedDataParallel":
+        r"""Applies ``fn`` recursively to every submodule (as returned by ``.children()``)
+        as well as self. Typical use includes initializing the parameters of a model
+        (see also :ref:`nn-init-doc`).
+
+        Compared to ``torch.nn.Module.apply``, this version additionally gathers
+        the full parameters before applying ``fn``. It should not be called from
+        within another ``summon_full_params`` context.
+
+        Args:
+            fn (:class:`Module` -> None): function to be applied to each submodule
+
+        Returns:
+            Module: self
+        """
+        uninitialized = self._is_root is None
+        self._assert_state(TrainingState_.IDLE)
+        with self._summon_full_params(recurse=False):
```

Comment:
hmm, the comment said 'Applies ``fn`` recursively to every submodule', but here _summon_full_params is called with recurse=False, not sure how it works

Reply:
It is similar to how `state_dict` will be implemented, we don't recurse as we call `super().apply()` which calls the method in nn.Module. That method should run `fn` on this module and then recursively call in on its children modules, so as a result the FSDP instances will have `apply` invoked so it makes sense for it to be nonrecursive.

Reply:
I see, that makes sense! Thanks for explaining